src/Catalyst.Core.Modules.Dfs/CoreApi/UnixFsApi.cs

```diff
@@ -0,0 +1,408 @@
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Catalyst.Abstractions.Dfs;
+using Catalyst.Abstractions.Dfs.CoreApi;
+using Catalyst.Abstractions.Keystore;
+using Catalyst.Abstractions.Options;
+using Catalyst.Core.Lib.Dag;
+using Catalyst.Core.Lib.IO;
+using Catalyst.Core.Modules.Dfs.UnixFs;
+using ICSharpCode.SharpZipLib.Tar;
+using Lib.P2P;
+using ProtoBuf;
+
+namespace Catalyst.Core.Modules.Dfs.CoreApi
+{
+    internal sealed partial class UnixFsApi : IUnixFsApi
+    {
+        private readonly IDhtApi _dhtApi;
+        private readonly IBlockApi _blockApi;
+        private readonly IKeyApi _keyApi;
+        private readonly INameApi _nameApi;
+        private readonly DfsState _dfsState;
+
+        private const int DefaultLinksPerBlock = 174;
+
+        public UnixFsApi(IDhtApi dhtApi, IBlockApi blockApi, IKeyApi keyApi, INameApi nameApi, DfsState dfsState)
+        {
+            _dhtApi = dhtApi;
+            _blockApi = blockApi;
+            _keyApi = keyApi;
+            _nameApi = nameApi;
+            _dfsState = dfsState;
+        }
+
+        public async Task<IFileSystemNode> AddFileAsync(string path,
+            AddFileOptions options = default,
+            CancellationToken cancel = default)
+        {
+            await using var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read);
+            return await AddAsync(stream, Path.GetFileName(path), options, cancel).ConfigureAwait(false);
+        }
+
+        public async Task<IFileSystemNode> AddTextAsync(string text,
+            AddFileOptions options = default,
+            CancellationToken cancel = default)
+        {
+            await using var ms = new MemoryStream(Encoding.UTF8.GetBytes(text), false);
+            return await AddAsync(ms, "", options, cancel).ConfigureAwait(false);
+        }
+
+        public async Task<IFileSystemNode> AddAsync(Stream stream,
+            string name,
+            AddFileOptions options,
+            CancellationToken cancel)
+        {
+            options ??= new AddFileOptions();
+
+            // TODO: various options
+            if (options.Trickle)
+            {
+                throw new NotImplementedException("Trickle");
+            }
+
+            var blockService = GetBlockService(options);
+
+            var chunker = new SizeChunker();
+            var nodes = await chunker.ChunkAsync(stream, name, options, blockService, _keyApi, cancel)
+               .ConfigureAwait(false);
+
+            // Multiple nodes for the file?
+            var node = await BuildTreeAsync(nodes, options, cancel);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)